### ISSUE TITLE
Loading Screen for Files

### ISSUE THREAD

**Issue by dupontinquiries** [pulsar-edit/pulsar#641]
### Have you checked for existing feature requests?

- [X] Completed

### Summary

When opening a file with 500-1000 lines of code, Pulsar's editor window can become unresponsive for a second and a half. This means that certain functionality such as scrolling is not available for a moment. I propose a concept for a loading phase or loading screen that solves this issue of Pulsar feeling unresponsive compared to once the window has loaded all extensions for the respective window.

### What benefits does this feature provide?

If the user were able to scroll through text while plugins were loading the editor would feel much more responsive. This would allow developers to read through a file and know if they opened the right one without having to wait 1 or 2 seconds for each time they switch tabs to a new file. It would also allow users to cancel the loading of plugins on a file if they decide to switch to the next available tab or open a new file. If a user is looking through their tabs and several are open or they have just opened multiple files, this penalty can stack and become 10-15 seconds of looking at a frozen GUI. Being used to a responsive editor most of the time but then having uncharacteristic hiccups like this can break train of thought.

Maybe there can also be a new event or css styling class called .fully-loaded that can signal to the stylesheet that the editor is loaded so that users can add fade-in animations or other custom styling that can make the editor feel more customizable.

### Any alternatives?

Trimming down on heavy extensions or asking the most significant extension authors to optimize within their own codebases would be a good alternative.

### Other examples:

A lot of applications have loading screens or file previews that are meant to load quickly.

**Comment by confused-Techie** [comment]
Thanks a ton for contributing and suggesting features!

I will say right off the bat, the freezing of loading is obviously not something we want to happen ever. It's more an issue with the render process of Electron becoming frozen as JavaScript is being executed, which doesn't have an easy way to solve. Meaning nothing can happen in the render process while this occurs.

Now there may be options we could look at, such as tokenizing the file or loading it within a web worker, but that introduces complex and difficult problems with deserializing complex data.

That's all to say this would be a fantastic feature, but implementation of such a thing may be extremely difficult and technically challenging.

**Comment by dupontinquiries** [comment]
Yes that seems very challenging. I guess you could mark this as closed unless someone wants to pick it up and work on an implementation.

**Comment by Cristianald00** [comment]
Hi there. I'm also experiencing the same issue and similar performance problems with frequent crashes in my Pulsar. Opening files is slow, impacting productivity, and sometimes the editor crashes, making it unusable.

I have only three external packages (colour-picker, file-icons, and language-vue), so they may not be the issue. However, I'm running numerous core packages (These were already active with the installation).

Any advice on improving Pulsar's performance would be greatly appreciated. I really like Pulsar so I am hoping to find a solution before considering a different editor.

**Comment by confused-Techie** [comment]
@Cristianald00 While this issue is about slow load times of files, it doesn't relate at all to Pulsar having frequent crashes. And that's not something we are seeing happening very often, or even being reported at all, outside a few issues.

I'd recommend creating your own issue about Pulsar having frequent crashes, as we don't expect to see that, so that we can try to find out what's going on.

---

As for improving Pulsar's performance, since I'll have to assume you are talking about when opening files, basically it's a limitation of the syntax highlighting, where the tools we use must highlight an entire file, so that if the file is extremely large it can take some time. The most common situation is with say, minified JavaScript, where an entire file is put onto one line. Our Syntax Highlighting works line by line, so if one line is super long, it can cause a warning saying the editor is unresponsive. If you say wait instead it gets there after a few moments, but takes a while. In those cases the best thing to do is:
* Either try to limit the length of any single line in your project
* Or what may be more practical is to disable syntax highlighting on files you know are going to be a problem, this way you can avoid the issue completely, even if you lose that feature on a file.

**Comment by Cristianald00** [comment]
@confused-Techie I'll try removing syntax highlighting and shortening the line length to address the problem. If these steps don't resolve it, I'll consider opening a new ticket with additional details. Thanks for the valuable information!

**Comment by savetheclocktower** [comment]
I can think of a couple of reasons why an upgrade to 1.131.1 might address this old issue:

1. We've received feedback from users on all platforms that Pulsar running on an upgraded Electron (we recently bumped from v12 to v30) resulted in across-the-board performance gains. The gains were most pronounced on startup, but I think it's quite possible that they'd extend to file opening as well.
2. Users on Apple Silicon in particular were singularly affected by an issue involving unreasonable slowness of spawning child processes. If a user had installed an IDE package or anything else where the opening of a file could've triggered side effects (like the starting of a language server), then it would've manifested as a one- or two-second freeze. In my experience, something as simple as focusing a new text editor had a chance of triggering a background `git` operation (usually a status check) that could lock up the editor for a second. This is finally fixed for Apple Silicon users on 1.131.

Since the reported issue specifically mentions files with at least 500 lines, it's possible that something else is going on. 500 lines isn't very much; there are a few files in the Pulsar codebase that are thousands of lines long, but which don't (in my experience) cause the editor to freeze even for a short while.

But @confused-Techie is right about extremely _long_ lines being Pulsar's kryptonite. There are two reasons I can think of for this:

1. When soft-wrapping is enabled, Pulsar's `text-buffer` library has to do extra work. Soft-wrapping makes it so that “buffer lines” and “screen lines” are different things, and must convert between the two coordinate systems in order to draw text accurately on screen. In files with lots of long lines, a lot of this work must be done. (Turning off “Soft Wrap” in the editor settings is one way to diagnose this; if performance improves, then buffer/screen coordinate conversions might be the bottleneck.)
2. Related to the above: Pulsar tries hard to apply syntax highlighting only to what the user can see on screen, rather than to the entire buffer. That's partly because the highlighting itself (no matter which system is used) does have to happen on the main “thread”; it's not its own job in a separate process or a web worker. Parsing for syntax highlighting is designed to go async, though. (Not familiar enough with the old TextMate grammar system to answer exactly how, but I can say that for Tree-sitter grammars, we enforce a max parsing budget of 3 milliseconds per frame; a frame is 16.67ms.) The system is designed so that, even if syntax highlighting takes a while to apply, the buffer should remain responsive and editable while it's happening.

    But there's a weakness in this system: Pulsar can highlight a subset of a buffer, but that subset must consist of some number of buffer lines. It's not currently possible to syntax-highlight only a fraction of a single line. So if a file has — god forbid — _one very long line_, it's the worst-case scenario, because it means that Pulsar must syntax-highlight the whole file at once.

   Addressing this is somewhat risky because we don't want to regress performance or functionality. But if you just want to know whether syntax highlighting itself is the culprit: save a copy of the offending file as a `.txt` file, then open it and see if you get the same freeze. Plaintext files open with a “null” grammar that does not attempt to do any syntax highlighting, so it's a good way to give your experiment a control group.

If this is still an issue for anyone, I'd like to suggest that they find a publicly available source file (something from any open-source repo) to demonstrate the issue — or else they can ask an LLM to write 3,000 lines of sample code in their favorite language. Then show it to us. I'd be curious to see whether there are any other potential bottlenecks at play, but I want to make sure we're working with input that triggers the slowness in the first place.
